File: MATLAB/functions/QUANTIFY.m
function baseline = QUANTIFY(params, fund, targets, options)
%QUANTIFY Match city moments and optional physical targets.
if nargin < 4, options = NUMERICS(); end % Use default numerical controls.
fields = {'floor_space_per_resident','residential_rent','total_urban_area','cbd_far'}; % Optional positive targets.
active = false(1,4); % Omitted fields preserve compatibility with older inputs.
for k = 1:numel(fields)
    active(k) = isfield(targets,fields{k}) && ~isempty(targets.(fields{k})); % Empty disables this moment.
    if active(k), validateattributes(targets.(fields{k}),{'numeric'},{'scalar','real','finite','positive'},mfilename,fields{k}); end % Reject invalid moments before solving.
end
if all(active(1:2)), error('ABSJ2027:HousingTargets','Choose either floor_space_per_resident or residential_rent, not both.'); end % Avoid conflicting housing targets.
if active(2) && (~isfield(targets,'wage') || isempty(targets.wage))
    error('ABSJ2027:RentUnits','A residential_rent target requires targets.wage: rent is annual currency per m^2, in the same currency as annual wages.'); % Define the monetary scale explicitly.
end
if active(3) && targets.total_urban_area > pi*min(options.radius,params.x1_max)^2
    error('ABSJ2027:AreaTarget','Total urban area exceeds the grid or urban growth boundary. The target is geographic km^2, including non-developable land.'); % Reject an impossible geographic target.
end
if ~isempty(targets.height_gap) && isinf(params.S_bar_C) && isinf(params.S_bar_R) && (any(active) || ~IS_PAPER_PARAMETERS(params))
    error('ABSJ2027:HeightGapParameterization','Height-gap calibration is only supported under the paper parameterization because T is calibrated for that model. Set targets.height_gap=[] for empirical quantification; the height gap is still reported as an outcome relative to T.'); % Do not fit a paper height gap after re-estimating physical parameters.
end
if active(4) && (params.tau_R<=0 || params.tau_C<=0)
    error('ABSJ2027:CBDTarget','CBD FAR calibration requires positive starting tau_R and tau_C to preserve their ratio.'); % A common scale cannot move a zero decay parameter.
end
if active(4) && isfinite(params.S_bar_C) && targets.cbd_far>=params.S_bar_C
    error('ABSJ2027:CBDTarget','The CBD FAR target must be below the manual commercial FAR cap; at the cap the tau scale is not identified.'); % Reject an impossible or capped central-height target.
end
if active(2) || active(3) || active(4)
    baseline = MATCH_CITY_TARGETS(params,fund,targets,options,active); % Jointly adjust construction costs and/or agricultural rent.
elseif active(1)
    baseline = MATCH_FLOORSPACE(params,fund,targets,options); % Preserve the existing floor-space-only calibration.
else
    baseline = QUANTIFY_CITY(params,fund,targets,options); % Paper baseline: retain supplied costs and agricultural rent.
end
baseline = ADD_HEIGHT_GAP(baseline); % Compute the reported outcome once, after all optional calibration has finished.

end

File: Python/functions/quantify.py
"""Public quantification dispatcher."""
import numpy as np
from add_height_gap import add_height_gap
from match_city_targets import match_city_targets
from quantify_city import quantify_city


def quantify(params, fund, targets, options):
    fields = ("floor_space_per_resident", "residential_rent", "total_urban_area", "cbd_far")
    active = [targets.get(name) is not None for name in fields]
    if active[0] and active[1]: raise ValueError("Choose either floor_space_per_resident or residential_rent, not both.")
    if (active[0] or active[1]) and not active[2]: raise ValueError("A housing quantity or rent target requires total_urban_area to identify physical space.")
    if active[1] and targets.get("wage") is None: raise ValueError("A residential_rent target requires an annual wage target.")
    if active[2] and targets["total_urban_area"] > np.pi*min(options["radius"], params["x1_max"])**2:
        raise ValueError("Total urban area exceeds the grid or urban growth boundary.")
    if targets.get("height_gap") is not None and np.isinf(params["S_bar_C"]) and np.isinf(params["S_bar_R"]) and any(active):
        raise ValueError("Height-gap calibration is only supported under the paper parameterization; omit it for empirical quantification.")
    baseline = match_city_targets(params, fund, targets, options, active) if any(active[1:]) else quantify_city(params, fund, options, targets)
    baseline["targets"] = dict(targets); baseline["options"] = dict(options)
    return add_height_gap(baseline, options, targets["urbanization_rate"])
